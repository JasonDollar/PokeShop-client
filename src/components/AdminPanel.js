import React, { useContext } from 'react'
import { useQuery } from '@apollo/react-hooks'
import gql from 'graphql-tag'
import { Redirect } from 'react-router-dom'
import { UserContext } from '../userContext'
import UserTableRow from './UserTableRow'
import WidthContainer from './styles/WidthContainer'
import Loading from './Loading'
import UsersTable from './styles/UsersTable'

export const USERS_ADMIN_QUERY = gql`
  query USERS_ADMIN_QUERY {
    users {
      id
      name
      email
      role
    }
  }
`



const AdminPanel = () => {
  const { userId } = useContext(UserContext)
  const { data, loading, error } = useQuery(USERS_ADMIN_QUERY, {
    variables: { adminId: userId },
  })
  if (!userId) return <Redirect to="/" />
  if (loading) return <Loading />
  if (error) return <p>{error.message}</p>
  return (
    <WidthContainer>
      <div>
        <UsersTable>
          <thead>
            <tr className="table-items">
              <th>ID</th>
              <th>Name</th>
              <th>Email</th>
              <th>Role</th>
              <th>
                <span role="img" aria-label="Actions">ðŸ”¥ðŸ”¥ðŸ”¥</span>
              </th>
            </tr>
          </thead>
          <tbody>
            {data.users.map(item => <UserTableRow key={item.id} user={item} />)}
          </tbody>
        </UsersTable> 
      </div>
    </WidthContainer>
  )
}

export default AdminPanel
