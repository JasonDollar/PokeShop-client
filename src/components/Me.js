import React from 'react'
import { useQuery, gql } from '@apollo/client'
import OfferListItem from './OfferListItem'
import WidthContainer from './styles/WidthContainer'
import GridList from './styles/GridList'
import UserInfo from './styles/UserInfo'
import Loading from './Loading'
import ErrorMessage from './ErrorMessage'
import { formatBigNumber } from '../utils'

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    me {
      id
      name
      email 
      role
      offers {
        id
        name
        price
        pokemon {
          id
          image
        }
      }
      wallet {
        balance
      }
    }
  }
`

const Me = () => {
  const { data, loading, error } = useQuery(CURRENT_USER_QUERY, {
    fetchPolicy: 'cache-and-network',
  })

  if (loading) return <Loading />
  if (error) return <ErrorMessage message={error.message} />
  if (!data.me) return <ErrorMessage message="We could not fetch your information. Check back later" />
  return (
    <WidthContainer>
      <UserInfo>
        <h2 className="user">{data.me.name}</h2>
        <h3 className="email">{data.me.email}</h3>
        {/* <p>{data.me.role}</p> */}
        <p className="balance">You have {formatBigNumber(data.me.wallet.balance)} CR in your wallet</p>
        <p className="selling">You are selling:</p>
        <GridList>
          {data.me.offers.map(item => (
            <OfferListItem key={item.id} pokemonOffer={item} />
          ))}
        </GridList>
      </UserInfo>
    </WidthContainer>
  )
}

export default Me
