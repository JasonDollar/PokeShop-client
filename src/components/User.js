import React from 'react'
import { Query } from 'react-apollo'
import gql from 'graphql-tag'
import PropTypes from 'prop-types'


export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    me {
      id
      name
      email 
      offers {
        id
        name
        price
        pokemon {
          id
          image
        }
      }
      wallet {
        balance
      }
    }
  }
`

const User = props => (
    <Query query={CURRENT_USER_QUERY} fetchPolicy="cache-and-network">
      {payload => props.children(payload)}
    </Query>
)

export default User

User.propTypes = {
  children: PropTypes.any.isRequired,
}