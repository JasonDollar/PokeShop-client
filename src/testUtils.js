const fakeWallet = () => ({
  __typename: 'Wallet',
  id: '123213',
  balance: 100000,
  owner: '4234',
})

const fakeWallet2 = () => ({
  __typename: 'Wallet',
  id: 'dfghdfghdfhg',
  balance: 100000,
  owner: '4255',
})

const fakeUser = () => ({
  __typename: 'User',
  id: '4234',
  name: 'Miss Hilda Fahey',
  email: 'Satterfield.Horacio@gmail.com',
  role: 'admin',
  offers: [],
  cart: [],
  wallet: fakeWallet(),
})

const fakeUser2 = () => ({
  __typename: 'User',
  id: '4255',
  name: 'Mrs. Xzavier Mueller',
  email: 'Miller_Reichert@yahoo.com',
  role: 'admin',
  offers: [],
  cart: [],
  wallet: fakeWallet(),
})

const fakePokemon = () => ({
  __typename: 'Pokemon',
  id: 1,
  name: 'bulbasaur',
  url: 'https://pokeapi.co/api/v2/pokemon/1/',
  image: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.pg',
  pokeType: ['grass'],
})

const fakePokemonOffer = () => ({
  __typename: 'PokemonOffer',
  id: '618726387213',
  name: 'bulbasaur',
  price: 1000,
  description: 'It is a fake pokemonOffer description',
  pokemon: fakePokemon(),
  seller: fakeUser(),
  createdAt: '2019-06-20T20:38:07.889+00:00',
})

const fakeCartItem = () => ({
  __typename: 'CartItem',
  id: '5d0d361471e9af00178faea2',
  quantity: 1,
  pokemon: fakePokemon(),
  user: fakeUser(),
})

const fakeOrderItem = (changes) => ({
  __typename: 'OrderItem',
  id: '5d0bef4b49cbfc0017dce192',
  quantity: 1,
  price: 1000,
  pokemon: fakePokemon(),
  user: fakeUser(),
  seller: fakeUser2(),
  ...changes,
})

const fakeOrder = () => ({
  __typename: 'Order',
  id: '5d0bef4b49cbfc0017dce193',
  price: 10000,
  user: fakeUser(),
  items: [fakeOrderItem(), fakeOrderItem({ id: '5d0bf3a549cbfc0017dce199' })],
})

export {
  fakeUser,
  fakeUser2,
  fakeWallet,
  fakePokemon,
  fakePokemonOffer,
  fakeCartItem,
  fakeOrderItem,
  fakeOrder,
}