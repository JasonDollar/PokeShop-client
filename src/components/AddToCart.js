import React from 'react'
import { useMutation, gql } from '@apollo/client'
import PropTypes from 'prop-types'
import { CART_ITEMS_QUERY } from './Cart'

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
      pokemon {
        id
        name
      }
    }
  }
`

const AddToCart = ({
  pokemonOfferId, children, disabledButton, CSSclass, 
}) => {
  const [addToCart, { loading }] = useMutation(ADD_TO_CART_MUTATION, {
    variables: { id: pokemonOfferId },
    refetchQueries: [{ query: CART_ITEMS_QUERY }],
  })

  return (
    <button 
      type="button"
      className={`${CSSclass} ${loading && 'loading'}`}
      onClick={addToCart}
      disabled={loading || disabledButton}
    >
      {children}
    </button>
  )
}

export default AddToCart

AddToCart.propTypes = {
  pokemonOfferId: PropTypes.string.isRequired,
  children: PropTypes.any,
  disabledButton: PropTypes.bool.isRequired,
  CSSclass: PropTypes.string,
}

AddToCart.defaultProps = {
  children: 'Add to Cart',
  CSSclass: '',
}