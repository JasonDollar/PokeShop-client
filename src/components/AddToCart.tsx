import React from 'react'
import { useMutation } from '@apollo/react-hooks'
import gql from 'graphql-tag'
import PropTypes from 'prop-types'

import { CART_ITEMS_QUERY } from './Cart'

interface Pokemon {
  id: number
  name: number
  url?: string
  image?:string
  type?: string[]
}

interface NewCartItem {
  id: string,
  quantity: number,
  pokemon: Pokemon
}

interface AddToCartProps {
  pokemonOfferId: string
  children: any
  disabledButton: boolean
  CSSclass: string
}

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
      pokemon {
        id
        name
      }
    }
  }
`

const AddToCart: React.FC<AddToCartProps> = ({
  pokemonOfferId, children, disabledButton, CSSclass, 
}): JSX.Element => {
  const [addToCart, { loading }] = useMutation<
    {addToCart: NewCartItem},
    {id: string}
  >(ADD_TO_CART_MUTATION, {
    variables: { id: pokemonOfferId },
    refetchQueries: [{ query: CART_ITEMS_QUERY }],
  })

  const onButtonClick = (): void => {
    addToCart()
  }

  return (
    <button 
      type="button"
      className={`${CSSclass} ${loading && 'loading'}`}
      onClick={onButtonClick}
      disabled={loading || disabledButton}
    >
      {children}
    </button>
  )
}

export default AddToCart

// AddToCart.propTypes = {
//   pokemonOfferId: PropTypes.string.isRequired,
//   children: PropTypes.any,
//   disabledButton: PropTypes.bool.isRequired,
//   CSSclass: PropTypes.string,
// }

// AddToCart.defaultProps = {
//   children: 'Add to Cart',
//   CSSclass: '',
// }