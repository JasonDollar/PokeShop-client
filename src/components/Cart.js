import React from 'react'
import { useQuery, gql } from '@apollo/client'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import RemoveFromCart from './RemoveFromCart'
import AddToCart from './AddToCart'
import Checkout from './Checkout'
import ShowMyMoney from './ShowMyMoney'
import Loading from './Loading'
import ErrorMessage from './ErrorMessage'
import { formatBigNumber } from '../utils'
import WidthContainer from './styles/WidthContainer'
import { EmptyCart, CartContainer, SinglePokemonCartItem } from './styles/Cart'

export const CART_ITEMS_QUERY = gql`
  query CART_ITEMS_QUERY {
    userCart {
      id
      quantity
      pokemon {
        id
        name
        price
        pokemon {
          id
          image
        }
      }
    }
  }
`

const Cart = () => {
  const { data, loading, error } = useQuery(CART_ITEMS_QUERY, {
    fetchPolicy: 'cache-and-network',
  })

  return (
    <WidthContainer>
      <ShowMyMoney>
        {({ data: moneyData, loading: moneyLoading }) => {
  
          if (loading || moneyLoading) return <Loading />
          if (error) return <ErrorMessage error={error} />
      
          if (data.userCart.length <= 0) {
            return <EmptyCart>Your cart is empty!</EmptyCart>
          }
          let totalPrice = 0
          if (data && data.userCart) {
            totalPrice = data.userCart.reduce((acc, item) => acc + (item.pokemon.price * item.quantity), 0)
          }
          const userCredits = moneyData.userCredits.balance
          const userHasEnoughCredits = userCredits > totalPrice
          return (
            <CartContainer>
              <ul>
                {data && data.userCart.map(item => (
                  <SinglePokemonCartItem key={item.id}>
                    <img src={item.pokemon.pokemon.image} alt={item.pokemon.name + ' image'} className="pokemonImage" />
                    <p className="pokemonText">
                      <span className="pokemonName">{item.pokemon.name}</span> - <strong>{formatBigNumber(item.pokemon.price)}</strong>CR each
                    </p>
                    <div className="cartControls">
                      <span className="quantityTitle">Quantity: </span>
                      <RemoveFromCart cartItemId={item.id} CSSclass="cartButton">
                        <FontAwesomeIcon icon="minus" />
                      </RemoveFromCart>
                      <div className="cartItemQuantity">{item.quantity}</div>
                      <AddToCart pokemonOfferId={item.pokemon.id} disabledButton={loading} CSSclass="cartButton">
                        <FontAwesomeIcon icon="plus" />
                      </AddToCart>
                    </div>
                    <p className="cartItemTotal">Total: <strong>{formatBigNumber(item.pokemon.price * item.quantity)}</strong>CR</p>
                  </SinglePokemonCartItem>
                ))}
              </ul>
              <p className="cartTotal">Total Price: {formatBigNumber(totalPrice)}CR</p>
              <Checkout buttonDisabled={!data || data.userCart.length <= 0 || !userHasEnoughCredits} totalPrice={totalPrice} />
              {!userHasEnoughCredits && <p className="overTheLimitError">You don&apos;t have enough credits! {formatBigNumber(totalPrice - userCredits)}CR over the limit.</p>}
            </CartContainer>
          )
        }}
      </ShowMyMoney>
    </WidthContainer>
  )
}


export default Cart
