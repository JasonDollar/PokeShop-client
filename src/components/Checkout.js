import React from 'react'
import { useMutation } from '@apollo/react-hooks'
import gql from 'graphql-tag'
import PropTypes from 'prop-types'
import { Redirect } from 'react-router-dom'
import styled from 'styled-components'
import { CART_ITEMS_QUERY } from './Cart'
import { USER_ORDERS_QUERY } from './OrderList'
import ShowMyMoney, { USER_CREDITS_QUERY } from './ShowMyMoney'
import ActionButton from './styles/ActionButton'

const ORDER_POKEMONS_MUTATION = gql`
  mutation ORDER_POKEMONS_MUTATION {
    orderPokemons  {
      id
      price
    }
  }
`

const CheckoutContainer = styled.div`
  text-align: center;
`


const Checkout = ({ buttonDisabled, totalPrice }) => {
  const [orderPokemons, { data, loading, error }] = useMutation(ORDER_POKEMONS_MUTATION, {
    refetchQueries: [
      { query: CART_ITEMS_QUERY }, 
      { query: USER_ORDERS_QUERY },
      { query: USER_CREDITS_QUERY },
    ],
  })

  return (
    <ShowMyMoney>
      {({ data: { userCredits }, loading: loadingCredits, error: errorCredits }) => {
        if (loadingCredits) return <p>Loading...</p>
        
        if (data) {
          return <Redirect to="/orders" />
        }
        return (
          <CheckoutContainer>
            <ActionButton type="button" onClick={orderPokemons} disabled={loading || loadingCredits || buttonDisabled || (userCredits && userCredits.balance < totalPrice)}>
              Buy these shiny pokemons!
            </ActionButton>
            {errorCredits && <p>{errorCredits.message}</p> }
            {error && <p> {error.message}</p>}
            
          </CheckoutContainer>
        )
      }}
    </ShowMyMoney>
  )
}


export default Checkout

Checkout.propTypes = {
  buttonDisabled: PropTypes.bool,
  totalPrice: PropTypes.number,
}

Checkout.defaultProps = {
  buttonDisabled: false,
  totalPrice: 0,
}