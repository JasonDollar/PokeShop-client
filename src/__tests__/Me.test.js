import React from 'react'
import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'
import wait from 'waait'
import { MockedProvider } from 'react-apollo/test-utils'
import { CURRENT_USER_QUERY } from '../components/User'
import Me from '../components/Me'
import { fakeUser } from '../testUtils'

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: fakeUser() } },
  },
]

const noDataMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: null } },
  },
]
const errorMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      errors: [{ message: 'Items Not Found!' }],
    },
  },
]

describe('<Me/>', () => {
  test('should render correctly when logged in', async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocks}>
        <Me />
      </MockedProvider>,
    )
    await wait(0)
    wrapper.update()
    expect(toJSON(wrapper.find('h2.user'))).toMatchSnapshot()
    expect(toJSON(wrapper.find('h3.email'))).toMatchSnapshot()
    expect(toJSON(wrapper.find('p.balance'))).toMatchSnapshot()
  })

  test('should render error', async () => {
    const wrapper = mount(
      <MockedProvider mocks={errorMocks}>
        <Me />
      </MockedProvider>,
    )
    await wait(0)
    wrapper.update()
    expect(wrapper.find('p').text()).toContain('Items Not Found!')
  })
  test('should present loading state', async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocks}>
        <Me />
      </MockedProvider>,
    )
    expect(wrapper.find('Loading').exists()).toBe(true)
  })
  test('should display correct information when no data is fetched', async () => {
    const wrapper = mount(
      <MockedProvider mocks={noDataMocks}>
        <Me />
      </MockedProvider>,
    )
    await wait(0)
    wrapper.update()
    expect(wrapper.find('p').text()).toContain('We could fetch your information. Check back later')
  })
})