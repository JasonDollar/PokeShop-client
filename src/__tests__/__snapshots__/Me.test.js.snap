// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Me/> should render correctly when logged in 1`] = `
<h2
  className="user"
>
  Miss Hilda Fahey
</h2>
`;

exports[`<Me/> should render correctly when logged in 2`] = `
<h3
  className="email"
>
  Satterfield.Horacio@gmail.com
</h3>
`;

exports[`<Me/> should render correctly when logged in 3`] = `
<p
  className="balance"
>
  You have 
  100,000
   CR in your wallet
</p>
`;
